### PR TITLE
V4 redis connection plugin

### PR DESCRIPTION
**Pull request by correiarmjoao**
### What are the changes and their implications?

Add connection-redis plugin and documentation.

## Checklist

- [x] Pull request is made to the "develop" branch
- [x] Tests added
- [x] Documentation added/updated
- [x] Code has been formatted with Prettier
- [ ] Edits from maintainers are allowed

### CODE REVIEW

**Review by Gervwyk — DISMISSED** [review]

File: packages/plugins/connections/connection-redis/package.json
Comment (SamTolmay) [inline packages/plugins/connections/connection-redis/package.json:22]: @correiarmjoao, please add yourself as a contributor here

File: packages/plugins/connections/connection-redis/package.json
Comment (SamTolmay) [inline packages/plugins/connections/connection-redis/package.json:45]: Please pin this dependency to a fixed version. We use pinned dependencies across the project. 
```
"redis": "4.0.1"
```

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.js
Comment (SamTolmay) [inline packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.js:45]: Maybe this will be more clear if the message reads:
```
Invalid parameters, command "${command}" parameters should be an array,
```

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.js
Comment (SamTolmay) [inline packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.js:63]: @Gervwyk In other requests we throw the original error to make it easier for developers to debug:
```js
try {
  // ...
} catch (error) {
  client.quit();
  throw error
}
```
I think we should do this here?

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js
Comment (SamTolmay) [inline packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js:57]: Please add tests with a mocked redis connection to test that the client and command functions receive the correct arguments, all the commands are converted to uppercase, errors like invalid command are thrown correctly etc.

This is a possible example of the mocks should look like, but I have not been able to test/run this so I am not sure that it will work:
```js
jest.mock('redis', () => {
  return {
    createClient: mockCreateClient,
  };
});

const mockCreateClient = jest.fn();

let mockCreateClient;

beforeAll(async () => {
  mockCreateClient = (await import('redis')).createClient;
});

test("mocked client", () => {
  const client = {
    on: jest.fn(),
    connect: jest.fn(),
    GET: jest.fn(),
  }
  mockCreateClient.mockImplementation(() => client);
  expect(mockCreateClient.mock.calls).toEqual(/*...*/);
  expect(client.connect.mock.calls).toEqual(/*...*/);
  expect(client.GET.mock.calls).toEqual(/*...*/);
})
```

**Review by SamTolmay — CHANGES_REQUESTED** [review]
Thanks @correiarmjoao, this is looking great! I am very happy with how nicely the library is integrated in a flexable way. Please have a look at my comments and add your feedback.

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.js
Comment (Gervwyk) [reply packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.js:63]: Yes, we always pass the original error. This should change, thanks.

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js
Comment (correiarmjoao) [reply packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js:212]: @SamTolmay The example works mock implementation works fine. 

Though I spent last Friday trying to make it work just to realise jest mocks were not working correctly.
It seems to be an issue with jest and swc, you can see it in [this thread](https://github.com/swc-project/jest/issues/14).

It was fixed with a patch [here](https://github.com/swc-project/jest/pull/44#issuecomment-980601069).

This issue is affecting the tests on the other plugins too, but for now I bumped the @swc/core in the connection-redis only and the mocks work correctly there.

Comment (SamTolmay) [reply packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js:212]: Okay thanks. I think we might have been running into this too. We will update the version for the other packages once we have merged this PR.

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js
Comment (SamTolmay) [inline packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js:68]: Please add a test for the case when the connection url is a string. (Line 22 in `Redis.js`)

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js
Comment (SamTolmay) [inline packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js:92]: We should also throw the error returned by the redis client and not a generic error in this case. Sorry, I missed it in the previous review. Could you please update that and then update this test?

**Review by SamTolmay — CHANGES_REQUESTED** [review]
This is looking great, the tests are nice and clear and easy to read. I have just requested a few small changes and then I am happy to merge 😀

**Comment by SamTolmay** [comment]
@correiarmjoao Please also merge the `v4` branch into this branch as we require the branch to be up to date with the base branch before we can merge a PR, and we cannot do this since the edits from maintainers are not allowed.

File: packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js
Comment (correiarmjoao) [reply packages/plugins/connections/connection-redis/src/connections/Redis/Redis/Redis.test.js]: @SamTolmay 
No worries, I updated the test and the code. In this case we will not need the try/catch and client.on() will throw it on the callback.

**Comment by correiarmjoao** [comment]
@SamTolmay
Made the changes and merged the v4 branch.
I found an issue regarding the "Allow edits from maintainers" now showing for me, it seems to be related with the fact that the merge request is being made from a repo that belongs to an organisation which I'm a member of, instead of a repo that is owned by me.

**Review by SamTolmay — APPROVED** [review]

**Review by Gervwyk — APPROVED** [review]

**Pull request merged**
